### PR TITLE
Update mvi response

### PR DESCRIPTION
**Pull request by kreek**
* added VBA corp id to MVI response (user.participant_id)
* refactored common mvi response to base class
* mvi response now handles nested name or id nodes

### CODE REVIEW

File: config/application.yml.example [hunk 1]
```diff
@@ -19,7 +19,7 @@ EVSS_SAMPLE_CLAIMANT_USER: '{"uuid": "1234", "first_name": "Jane","last_name": "
 MVI_URL: 'http://ps-dev.commserv.healthevet.va.gov:8110/psim_webservice/IdMWebService'
 VHA_MAPSERVER_URL: 'https://maps.va.gov/server/rest/services/PROJECTS/Facility_Locator/MapServer'
 VHA_MAPSERVER_LAYER: '0'
-MOCK_MVI_SERVICE: true
+MOCK_MVI_SERVICE: 'true'
```
Comment (omgitsbillryan) [inline config/application.yml.example]: Why the change?

Comment (kreek) [inline config/application.yml.example]: env vars come through as strings and figaro gives a warning if you don't wrap in quotes

**Comment by robbiethegeek** [comment]
@carlmjohnson @austin-martinez @omgitsbillryan added you folks to assignee, please review, would love to have this merged ASAP

File: lib/mvi/responses/base.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        success: 'AA',
+        failure: 'AE',
+        invalid_request: 'AR'
+      }.freeze
+
+      class << self
+        attr_accessor :endpoint
+      end
+
+      def self.mvi_endpoint(endpoint)
```
Comment (omgitsbillryan) [inline lib/mvi/responses/base.rb]: Why does this exist?  It provides a way to set the Class variable `endpoint` which the code before it already does.
```ruby
class << self
  attr_accessor :endpoint
end
```

File: lib/mvi/responses/base.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        @endpoint = endpoint
+      end
+      delegate :endpoint, to: 'self.class'
```
Comment (omgitsbillryan) [inline lib/mvi/responses/base.rb]: I worry this could be a 'gotcha' for someone handling an instance of a class that extends `Base`.  For example,
```ruby
Foo.new.endpoint = 'my_endpoint'
```
without looking at `MVI::Responses:Base`, I would automatically assume `endpoint` to be an instance variable.

File: lib/mvi/responses/base.rb (hunk 2 above)
Comment (austin-martinez) [inline lib/mvi/responses/base.rb]: Agreed with @omgitsbillryan This creates a DSL for retrieving the right section of data from the response, but it is a bit more confusing that it needs to be. Could this just be a property of the class and the subclasses can override it?

File: spec/lib/mvi/responses/find_candidate_spec.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            mhv_id: '123456^PI^200MHV^USVHA^A',
+            ssn: '555443333',
+            status: 'active'
+          )
+        end
+      end
+
+      context 'when name parsing fails' do
+        it 'should set the names to false' do
+          allow_any_instance_of(String).to receive(:capitalize).and_raise(StandardError)
```
Comment (austin-martinez) [inline spec/lib/mvi/responses/find_candidate_spec.rb]: This test is a bit confusing, if a response causes name parsing to fail, is it still valid? Stubbing String to raise an error on capitalize does not mimic a real scenario, can we use data that would cause the name parsing to fail (something that would call capitalize on nil or similar)?

File: lib/mvi/responses/find_candidate.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+require_relative 'base'
+
+module MVI
+  module Responses
+    class FindCandidate < Base
```
Comment (omgitsbillryan) [inline lib/mvi/responses/find_candidate.rb]: Any chance we can some informal doc here in the form of a comment?  Looks like it's responsible for parsing the MVI response but it's unclear to me what form the MVI Response is in.

File: lib/mvi/responses/find_candidate.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        ssn_id.first.dig(:id, :@extension)
+      rescue => e
+        Rails.logger.warn "MVI::Response.parse_ssn failed: #{e.message}"
+        nil
+      end
+
+      # MVI correlation id source id relationships:
+      # {source id}^{id type}^{assigning authority}^{assigning facility}^{id status}
+      # NI = national identifier, PI = patient identifier
+      def map_correlation_ids(ids)
```
Comment (omgitsbillryan) [inline lib/mvi/responses/find_candidate.rb]: I don't like large overbearing comments - but it feels warranted here.  This method may as well be written in hieroglyphics - I have no clue what's going on.

File: lib/mvi/responses/find_candidate.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        { given: given, family: family }
+      rescue => e
+        Rails.logger.warn "MVI::Response.parse_name failed: #{e.message}"
+        { given: nil, family: nil }
+      end
+
+      # other_ids can be hash or array of hashes
+      def parse_ssn(other_ids)
+        other_ids = [other_ids] if other_ids.is_a? Hash
+        ssn_id = other_ids.select { |id| id.dig(:id, :@root) == '2.16.840.1.113883.4.1' }
```
Comment (austin-martinez) [inline lib/mvi/responses/find_candidate.rb]: Should this be a constant?

File: lib/mvi/responses/find_candidate.rb [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+      # MVI correlation id source id relationships:
+      # {source id}^{id type}^{assigning authority}^{assigning facility}^{id status}
+      # NI = national identifier, PI = patient identifier
+      def map_correlation_ids(ids)
+        {
+          icn: select_extension(ids, /^\w+\^NI\^\w+\^\w+\^\w+$/, '2.16.840.1.113883.4.349'),
+          mhv_id: select_extension(ids, /^\w+\^PI\^200MHV\^\w+\^\w+$/, '2.16.840.1.113883.4.349'),
+          edipi: select_extension(ids, /^\w+\^NI\^200DOD\^USDOD\^\w+$/, '2.16.840.1.113883.3.364'),
+          vba_corp_id: select_extension(ids, /^\w+\^PI\^200CORP\^USVBA\^\w+$/, '2.16.840.1.113883.4.349')
```
Comment (austin-martinez) [inline lib/mvi/responses/find_candidate.rb]: Can the last argument be a constant? Looks like 3 are the same, is that a coincidence or would a change to one necessitate a change to the others?

File: spec/lib/mvi/responses/find_candidate_spec.rb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        )
+      end
+    end
+  end
+
+  context 'given an invalid response' do
+    let(:invalid_response) { MVI::Responses::FindCandidate.new(mvi_savon_invalid_response) }
+
+    describe '#invalid?' do
+      it 'should return false' do
```
Comment (austin-martinez) [inline spec/lib/mvi/responses/find_candidate_spec.rb]: This description doesn't match the spec

File: spec/lib/mvi/responses/find_candidate_spec.rb [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    let(:failure_response) { MVI::Responses::FindCandidate.new(mvi_savon_failure_response) }
+
+    describe '#invalid?' do
+      it 'should return false' do
+        expect(failure_response.invalid?).to be_falsey
+      end
+    end
+
+    describe '#failure?' do
+      it 'should return false' do
```
Comment (austin-martinez) [inline spec/lib/mvi/responses/find_candidate_spec.rb]: Same as above, description should match spec

File: lib/mvi/responses/find_candidate.rb (hunk 8 above)
Comment (kreek) [inline lib/mvi/responses/find_candidate.rb]: 👍 it's MVI's way of defining the xml node type so they won't change.

File: lib/mvi/responses/base.rb (hunk 2 above)
Comment (kreek) [inline lib/mvi/responses/base.rb]: @omgitsbillryan the idea was to make subclass config DSL like, @austin-martinez I agree it does make the base class more complex, we have a similar pattern for the redis store.

File: spec/support/mvi/find_candidate_response.xml [hunk 11]
```diff
@@ -50,6 +50,7 @@
                 <id extension="TKIP123456^PI^200IP^USVHA^A" root="2.16.840.1.113883.4.349"/>
                 <id extension="123456^PI^200MHV^USVHA^A" root="2.16.840.1.113883.4.349"/>
                 <id extension="1234^NI^200DOD^USDOD^A" root="2.16.840.1.113883.3.364" />
+                <id extension="12345678^PI^200CORP^USVBA^A" root="2.16.840.1.113883.4.349"/>
```
Comment (saneshark) [inline spec/support/mvi/find_candidate_response.xml]: We should look at putting all of these in a consistent place. For MHV we used `spec/support/fixtures` and I made that decision mostly because I felt `spec/fixtures` is more for transactional fixtures related to active_record. I guess it doesn't really matter, but we should put them all in a consistent place.
